src/lib/operators/join_mpsm/column_materializer_numa.hpp

```diff
@@ -0,0 +1,231 @@
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "resolve_type.hpp"
+#include "scheduler/current_scheduler.hpp"
+#include "scheduler/job_task.hpp"
+#include "storage/create_iterable_from_column.hpp"
+#include "storage/numa_placement_manager.hpp"
+#include "types.hpp"
+#include "utils/numa_memory_resource.hpp"
+
+namespace opossum {
+
+template <typename T>
+struct MaterializedValue {
+  MaterializedValue() = default;
+  MaterializedValue(RowID row, T v) : row_id{row}, value{v} {}
+
+  RowID row_id;
+  T value;
+};
+
+template <typename T>
+using MaterializedValueAllocator = PolymorphicAllocator<MaterializedValue<T>>;
+
+template <typename T>
+using MaterializedChunk = std::vector<MaterializedValue<T>, MaterializedValueAllocator<T>>;
+
+template <typename T>
+struct MaterializedNUMAPartition {
+  NodeID _node_id;
+  MaterializedValueAllocator<T> _alloc;
+  std::vector<std::shared_ptr<MaterializedChunk<T>>> _chunk_columns;
+
+  explicit MaterializedNUMAPartition(NodeID node_id, size_t reserve_size)
+      : _node_id{node_id},
+        _alloc{NUMAPlacementManager::get().get_memory_resource(node_id)},
+        _chunk_columns(reserve_size) {}
+
+  MaterializedNUMAPartition() {}
+
+  void fit() {
+    _chunk_columns.erase(
+        std::remove(_chunk_columns.begin(), _chunk_columns.end(), std::shared_ptr<MaterializedChunk<T>>{}),
+        _chunk_columns.end());
+  }
+};
+
+template <typename T>
+using MaterializedNUMAPartitionList = std::vector<MaterializedNUMAPartition<T>>;
+
+/**
+ * Materializes a table for a specific column and sorts it if required. Row-Ids are kept in order to enable
+ * the construction of pos lists for the algorithms that are using this class.
+ **/
+template <typename T>
+class ColumnMaterializer {
```

Comment:
Is it necessary to have a second column_materializer instead of unifying both?

Or is it planned to merge them anyways?

Reply:
While they share a lot of the core semantics, our materializer has a different shape for the returned data and introduces overhead that is unneeded for the non-numa case. Currently there is no plan for merging them, especially since they are well isolated from the rest of the system.

Reply:
Ok. Could you please document what your materializer does on top of "simply materializing" so that is easier to get why there are do materializers?